platform/linux-generic/odp_packet_io.c

```diff
@@ -3242,66 +3250,95 @@ static int pktout_send_lso(odp_pktout_queue_t queue, odp_packet_t packet,
 		offset = hdr_len + (i * payload_len);
 		if (odp_packet_copy_from_pkt(pkt_out[i], hdr_len, packet, offset, payload_len)) {
 			ODP_ERR("Payload copy failed\n");
-			num_free = num_pkt;
 			goto error;
 		}
 	}
 
-	if (left_over) {
+	/* Copy left over payload */
+	if (left_over_len) {
 		offset = hdr_len + (num_full * payload_len);
 		if (odp_packet_copy_from_pkt(pkt_out[num_pkt - 1], hdr_len, packet, offset,
 					     left_over_len)){
 			ODP_ERR("Payload copy failed\n");
-			num_free = num_pkt;
 			goto error;
 		}
 	}
 
-	if (lso_proto == ODP_LSO_PROTO_IPV4) {
-		l3_offset = odp_packet_l3_offset(packet);
+	if (lso_prof->param.lso_proto == ODP_LSO_PROTO_IPV4) {
+		offset = odp_packet_l3_offset(packet);
 
-		if (l3_offset == ODP_PACKET_OFFSET_INVALID) {
+		if (offset == ODP_PACKET_OFFSET_INVALID) {
 			ODP_ERR("Invalid L3 offset\n");
-			num_free = num_pkt;
 			goto error;
 		}
 
 		for (i = 0; i < num_pkt; i++) {
-			if (lso_update_ipv4(pkt_out[i], i, num_pkt, l3_offset, payload_len)) {
+			if (lso_update_ipv4(pkt_out[i], i, num_pkt, offset, payload_len)) {
 				ODP_ERR("IPv4 header update failed. Packet %i.\n", i);
-				num_free = num_pkt;
 				goto error;
 			}
 		}
 	} else {
 		/* Update custom fields */
+		int num_custom = lso_prof->param.custom.num_custom;
+
 		for (i = 0; num_custom && i < num_pkt; i++) {
 			if (lso_update_custom(lso_prof, pkt_out[i], i)) {
 				ODP_ERR("Custom field update failed. Segment %i\n", i);
-				num_free = num_pkt;
 				goto error;
 			}
 		}
 	}
 
+	return 0;
+
+error:
+	odp_packet_free_multi(pkt_out, num_free);
+	return -1;
+}
+
+static int pktout_send_lso(odp_pktout_queue_t queue, odp_packet_t packet,
+			   const odp_packet_lso_opt_t *lso_opt)
+{
+	int ret, num_pkt;
+	uint32_t payload_len, left_over_len;
+
+	/* Calculate number of packets */
+	num_pkt = _odp_lso_num_packets(packet, lso_opt, &payload_len, &left_over_len);
+	if (odp_unlikely(num_pkt <= 0))
+		return -1;
+
+	/* Create packets */
+	odp_packet_t pkt_out[num_pkt];
+
+	ret = _odp_lso_create_packets(packet, lso_opt, payload_len, left_over_len, pkt_out,
+				      num_pkt);
+
+	if (odp_unlikely(ret))
+		return -1;
+
+	/* Send LSO packets */
 	ret = odp_pktout_send(queue, pkt_out, num_pkt);
 
 	if (ret < num_pkt) {
+		int first_free = 0;
+		int num_free = num_pkt;
+
 		ODP_DBG("Packet send failed %i\n", ret);
-		num_free = num_pkt;
+
 		if (ret > 0) {
 			first_free = ret;
 			num_free = num_pkt - ret;
 		}
+
+		odp_packet_free_multi(&pkt_out[first_free], num_free);
+		return -1;
 	}
```

Comment:
Perhaps this could be simplified to:
```C
if (ret < num_pkt) {
    const int cnt = ret >= 0 ? ret : 0; // Could also use MAX() macro defined in odp_macros_internal.h
    ODP_DBG("Packet send failed %i\n", ret);
    odp_packet_free_multi(&pkt_out[cnt], num_pkt - cnt);
    return -1;
}
```